src/crimson/osd/osd.cc

```diff
@@ -1114,6 +1132,82 @@ seastar::future<> OSD::_handle_osd_map(Ref<MOSDMap> m)
   });
 }
 
+/*
+ * Compare between the previous last_map we had to
+ * each one of the added_maps.
+ * Track all of the changes relevant in pg_num_history.
+ */
+seastar::future<> OSD::track_pools_and_pg_num_changes(
+  const std::map<epoch_t, OSDMapService::local_cached_map_t>& added_maps,
+  ceph::os::Transaction& t)
+{
+  LOG_PREFIX(OSD::track_pools_and_pg_num_changes);
+  epoch_t first = added_maps.begin()->first;
+  epoch_t last = added_maps.rbegin()->first;
+
+  // Unless this is the first start of this OSD,
+  // lastmap should be the newest_map we have.
+  OSDMapService::local_cached_map_t lastmap;
+
+  if (superblock.maps.empty()) {
+    DEBUG(" no maps stored, this is probably the first start of this osd");
+    lastmap = added_maps.at(first);
+  } else {
+    if (first > superblock.get_newest_map() + 1) {
+      ceph_assert(first == superblock.cluster_osdmap_trim_lower_bound);
+      DEBUG(" can't get previous map {} first start of this osd after a map gap",
+	  superblock.get_newest_map());
+    }
+  }
+  return pg_shard_manager.get_local_map(
+    superblock.get_newest_map()
+  ).then([this, FNAME, &t, added_maps, last](auto lastmap) {
+    for (auto& [current_added_map_epoch, current_added_map] : added_maps) {
+      _track_pools_and_pg_num_changes(t, lastmap,
+                                    current_added_map,
+                                    current_added_map_epoch);
+      lastmap = current_added_map;
+    }
+    pg_num_history.epoch = last;
+    return seastar::now();
+  });
+}
+
+void OSD::_track_pools_and_pg_num_changes(
+  ceph::os::Transaction& t,
+  const OSDMapService::local_cached_map_t& lastmap,
+  const OSDMapService::local_cached_map_t& current_added_map,
+  epoch_t current_added_map_epoch)
```

Comment:
out of scope: most of the logic here is identical to classic, we should consider porting it to a common helper.